### PR TITLE
Feat/lens follower substream changes

### PR DESCRIPTION
**Pull request by gopi-bathala**

### CODE REVIEW

File: lens-follower-tracker/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+use substreams::errors::Error;
+use substreams::{Hex};
+use substreams_ethereum::{pb::eth::v2 as eth, Event as EventTrait};
+use helpers::utils::format_with_0x;
+
+
+// Lens LPP contract 
+// const LENS_LPP_CONTRACT : &str = "db46d1dc155634fbc732f92e853b10b288ad5a1d";
+const ZERO_ADDRESS : &str = "0x0000000000000000000000000000000000000000";
+const LENS_PROXY_ACTION_ADDRESS : &str = "0xdd3f6c22ecc68007cc9f76da18984995da4b7b82";
```
Comment (0xsarvesh) [inline lens-follower-tracker/src/lib.rs]: Lets try injecting this from yaml file.

File: lens-follower-tracker/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+use pb::lens::events::{FollowNftDeployed, FollowEvents, FollowNftTransferred};
+use substreams::errors::Error;
+use substreams::{Hex};
+use substreams_ethereum::{pb::eth::v2 as eth, Event as EventTrait};
+use helpers::utils::format_with_0x;
+
+
+// Lens LPP contract 
+// const LENS_LPP_CONTRACT : &str = "db46d1dc155634fbc732f92e853b10b288ad5a1d";
+const ZERO_ADDRESS : &str = "0x0000000000000000000000000000000000000000";
```
Comment (0xsarvesh) [inline lens-follower-tracker/src/lib.rs]: Move this to constants or utils.

File: lens-follower-tracker/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        if !contract_address.contains(&addres) {
+           continue;
+        }
+        // for FollowNftDeployed event
+        if let Some(event) = abis::lens_events::events::FollowNftDeployed::match_and_decode(log) {
+            let follow_nft_deployed_event : FollowNftDeployed = FollowNftDeployed {
+                follow_profile_id: event.profile_id.to_string(),
+                follow_token_address: helpers::utils::format_with_0x(Hex(event.follow_nft).to_string())
+            };
+            follow_nft_deployed_events.push(follow_nft_deployed_event)
```
Comment (0xsarvesh) [inline lens-follower-tracker/src/lib.rs]: Execution can continue from this point. No need to decode other events.

File: lens-follower-tracker/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+        // for Followed event
+        if let Some(event) = abis::lens_events::events::FollowNftTransferred::match_and_decode(log) {
+            
+            let from = format_with_0x(Hex(event.from).to_string());
+            let to = format_with_0x(Hex(event.to).to_string());
+            
+            // scenario 1: follow event [from:0x to: follower address]
+            if from == ZERO_ADDRESS && to != LENS_PROXY_ACTION_ADDRESS {
+                let follow_nft_transferred_event : FollowNftTransferred = FollowNftTransferred {
```
Comment (0xsarvesh) [inline lens-follower-tracker/src/lib.rs]: Lets capture txHash, logIndex & blockNumber also.

File: lens-follower-tracker/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    for log in blk.logs() {
+        let addres : String = Hex(log.address().clone()).to_string();
+        if !contract_address.contains(&addres) {
+           continue;
+        }
+        // for FollowNftDeployed event
+        if let Some(event) = abis::lens_events::events::FollowNftDeployed::match_and_decode(log) {
+            let follow_nft_deployed_event : FollowNftDeployed = FollowNftDeployed {
+                follow_profile_id: event.profile_id.to_string(),
```
Comment (0xsarvesh) [inline lens-follower-tracker/src/lib.rs:42]: Lets capture txHash, logIndex & blockNumber also.

File: lens-follower-tracker/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                let follow_nft_transferred_event : FollowNftTransferred = FollowNftTransferred {
+                    follow_profile_id: event.profile_id.to_string(),
+                    follow_token_id: event.follow_nft_id.to_string(),
+                    follower_address: to.clone(),
+                    follow_type: FOLLOW.to_string(),
+                };
+                follow_nft_transferred_events.push(follow_nft_transferred_event) 
+            }
+
+            // scenario 2: follow for event [from:0x..782 [proxy] to: follower address]
```
Comment (0xsarvesh) [inline lens-follower-tracker/src/lib.rs:74]: Its good, if you can put sample txHash in comment

**Review by 0xsarvesh — CHANGES_REQUESTED** [review]
Some feedbacks & comments

File: lens-follower-tracker/src/lib.rs (hunk 1 above)
Comment (gopi-bathala) [reply lens-follower-tracker/src/lib.rs]: done

File: lens-follower-tracker/src/lib.rs (hunk 2 above)
Comment (gopi-bathala) [reply lens-follower-tracker/src/lib.rs]: done

File: lens-follower-tracker/src/lib.rs (hunk 4 above)
Comment (gopi-bathala) [reply lens-follower-tracker/src/lib.rs]: done

File: lens-follower-tracker/src/lib.rs (hunk 5 above)
Comment (gopi-bathala) [reply lens-follower-tracker/src/lib.rs:42]: done

File: lens-follower-tracker/src/lib.rs (hunk 6 above)
Comment (gopi-bathala) [reply lens-follower-tracker/src/lib.rs:74]: done

File: lens-follower-tracker/src/lib.rs (hunk 3 above)
Comment (gopi-bathala) [reply lens-follower-tracker/src/lib.rs]: done

**Review by 0xsarvesh — APPROVED** [review]

**Pull request merged**